### ISSUE TITLE
We should use mypy for better reliability and security

### ISSUE THREAD

**Issue by abompard** [fedora-infra/noggin#69]
[Mypy](http://mypy-lang.org/) is a static type checker for python. I think we should adopt it in securitas because it will detect a whole category of bugs for us, and thus improve reliability of the app.

Mypy has been used successfully in parts of bodhi, and works best when all the source code adopts it (as opposed to just parts of it). For this reason, I think it's best to introduce it as early as possible.

**Comment by relrod** [comment]
I'm almost always +1 on static typing, though I think "optional" (gradual) type checkers like mypy come with their own class of problems. Somewhat amusingly, Mypy claims to have a "Powerful type system" on its landing page, yet lacks some of the most basic constructions in all of type theory (such as closed sum types).

Although I love static typing and study type theory, logic, and category theory as a hobby, I'm +0 on this with the following considerations:

* If we're okay with using static typing, as you said, I'd want to use it everywhere. If we make exceptions, we immediately lose 100% of the benefits. This means avoiding escape hatches such as the top (⊤) type (`Any`), type casting (unsound), etc. This takes a lot of social commitment to uphold and is particularly hard in a language like Python where people are not used to working with type systems.

* I'd almost suggest that if we are going to do that, we should just look at using a language that was built around types from the ground up (Haskell or Ocaml, for example), rather than using a language which has tacked them on later. The advantage is all sorts of extra safety and type-level features. The disadvantage (and why I didn't write Securitas in Haskell -- my go-to language of choice) is that nobody on the team knows the language.

Basically I won't whine if we decide to use a tacked on static type system, but I *will* whine if we use a tacked on type system "except for this file" or "except for this function" or "except for this line" or "except for this type-cast". If we do this, I'd want to see us go all in on it. Otherwise it's just theater and cruft.

**Comment by abompard** [comment]
I think I understand. However using mypy in parts of the bodhi code did discover some bugs that were lurking and not covered by unit tests, so even if it does not provide as much benefit as using it everywhere, it is not entirely useless either, no?

I wonder what the rest of the team thinks, what the experiences are with static typing in python, if the benefits are obvious enough to everyone that we want to commit to this.

**Comment by StephenCoady** [comment]
I don't have the academic interest that @relrod does, but I do heavily favor typed code. I am also definitely in favor of introducing type-checking as early as possible. I do have a question though - what would be the reason for only introducing it in certain parts of the codebase? Is it a python-specific nuance that I'm not aware of?

**Comment by abompard** [comment]
> I do have a question though - what would be the reason for only introducing it in certain parts of the codebase? Is it a python-specific nuance that I'm not aware of?

I'm not sure, maybe this would not be as necessary in the unit tests? I've only used mypy on bodhi though, so I may be wrong.

**Comment by sfinn85** [comment]
System to help detect bugs before they occur. Prob doesn't make sense to after, alot of work - back to code. Could of done this first - lessons learned.
